aws_crt_common_ws/aws-crt-c-flags/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+use cc::Build;
+use fs_extra;
+use fs_extra::dir::CopyOptions;
+use gag::Gag;
+use serde::{Deserialize, Serialize};
+use serde_json::Result;
+use std::env;
+use std::fs;
+use std::path::Path;
+use std::path::PathBuf;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CRTModuleBuildInfo {
+    crt_module_name: String,
+    crt_module_deps: Vec<CRTModuleBuildInfo>,
+    private_cflags: Vec<String>,
+    public_cflags: Vec<String>,
+    private_defines: Vec<(String, String)>,
+    public_defines: Vec<(String, String)>,
+    link_targets: Vec<String>,
+    shared_lib: bool,
+    lib_name: String,
+    linker_path: Option<PathBuf>,
+    include_dirs: Vec<PathBuf>,
+    #[serde(skip_serializing, skip_deserializing)]
+    build_toolchain: Build,
+}
+
+impl CRTModuleBuildInfo {
+    /// Creates a new instance of CRTModuleBuildInfo.
+    /// # Arguments
+    ///
+    /// * `module_name` - Name of the module you want to build. This name will be used to identify
+    ///                   build state across crates. We recommend you name it your sys crate name.
+    ///                   For example: aws_crt_common_sys, aws_crt_http_sys etc...
+    ///
+    /// # Examples
+    ///
+    /// ```should_panic
+    /// use aws_crt_c_flags::{CRTModuleBuildInfo};
+    /// let build_info = CRTModuleBuildInfo::new("aws_crt_common_sys");
+    /// ```
+    pub fn new(module_name: &str) -> CRTModuleBuildInfo {
+        CRTModuleBuildInfo {
+            crt_module_name: module_name.parse().unwrap(),
+            crt_module_deps: vec![],
+            private_cflags: vec![],
+            public_cflags: vec![],
+            private_defines: vec![],
+            public_defines: vec![],
+            link_targets: vec![],
+            shared_lib: false,
+            lib_name: module_name.parse().unwrap(),
+            linker_path: Option::from(PathBuf::from(env::var_os("OUT_DIR").unwrap())),
+            include_dirs: vec![],
+            build_toolchain: Build::new(),
+        }
+    }
+
+    /// Declares other aws crt modules to have your sys package depend on. This is used for transitively
+    /// passing linker arguments and c-flags between different crates' builds of their sys modules.
+    ///
+    /// # Arguments
+    ///
+    /// * `dependency` - name of the crt sys crate you want to link your sys crate against. So for example,
+    ///                  if you're building aws-checksums, your sys crate would be aws_crt_checksums_sys, and you
+    ///                  would declare a dependency here on aws_crt_common_sys.
+    ///
+    /// # Examples
+    /// ```should_panic
+    /// use aws_crt_c_flags::{CRTModuleBuildInfo};
+    /// let mut build_info = CRTModuleBuildInfo::new("aws_crt_checksums_sys");
+    /// build_info.add_module_dependency("aws_crt_common_sys");
+    /// ```
+    pub fn add_module_dependency(&mut self, dependency: &str) -> &mut CRTModuleBuildInfo {
```

Comment:
This looks like a builder, and @rcoh 's feedback seems to echo that, if you remove all the `add_`, this should be `with_dependency` or similar